Description: set correct level on syslog messages
Origin: http://www.vdr-portal.de/board17-developer/board97-vdr-core/p1148688-/#post1148688
Forwarded: no
Author: Lars Hanisch <dvb@flensrocker.de>

--- a/tools.h
+++ b/tools.h
@@ -32,8 +32,8 @@
 extern int SysLogLevel;
 
 #define esyslog(a...) void( (SysLogLevel > 0) ? syslog_with_tid(LOG_ERR, a) : void() )
-#define isyslog(a...) void( (SysLogLevel > 1) ? syslog_with_tid(LOG_ERR, a) : void() )
-#define dsyslog(a...) void( (SysLogLevel > 2) ? syslog_with_tid(LOG_ERR, a) : void() )
+#define isyslog(a...) void( (SysLogLevel > 1) ? syslog_with_tid(SysLogLevel > 3 ? LOG_ERR : LOG_INFO, a) : void() )
+#define dsyslog(a...) void( (SysLogLevel > 2) ? syslog_with_tid(SysLogLevel > 3 ? LOG_ERR : LOG_DEBUG, a) : void() )
 
 #define LOG_ERROR         esyslog("ERROR (%s,%d): %m", __FILE__, __LINE__)
 #define LOG_ERROR_STR(s)  esyslog("ERROR (%s,%d): %s: %m", __FILE__, __LINE__, s)
--- a/vdr.c
+++ b/vdr.c
@@ -371,7 +371,7 @@
                        *p = 0;
                     if (isnumber(optarg)) {
                        int l = atoi(optarg);
-                       if (0 <= l && l <= 3) {
+                       if (0 <= l && l <= 4) {
                           SysLogLevel = l;
                           if (!p)
                              break;
